### PR TITLE
Edge devices create - feature tweaks and tests

### PR DESCRIPTION
**Pull request by c-ryan-k**
Only set auto-reprovision mode to Dynamic if not set

Edit existing device-identity error message, add tests

Added test for file write error handling

---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

Intent for Production

- [x] It is expected that pull requests made to default or core branches such as `dev` or `main` are of production grade. Corollary to this, any merged contributions to these branches may be deployed in a public release at any given time. By checking this box, you agree and commit to the expected production quality of code.

Basic expectations

- [x] If introducing new functionality or modified behavior, are they backed by unit and/or integration tests?
- [x] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** the relevant unit **and** integration tests pass? i.e. `pytest <project root> -vv`. Please provide evidence in the form of a screenshot showing a succesful run of tests locally OR a link to a test pipeline that has been run against the change-set.
- [x] Have linter checks passed using the `.pylintrc` and `.flake8` rules? Look at the CI scripts for example usage.
- [x] Have extraneous print or debug statements, commented out code-blocks or code-statements (if any) been removed from the surface area of changes?
- [ ] Have you made an entry in HISTORY.rst which concisely explains your user-facing feature or change?

Azure IoT CLI maintainers reserve the right to enforce any of the outlined expectations.

A PR is considered **ready for review** when all basic expectations have been met (or do not apply).

### CODE REVIEW

File: azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py [hunk 1]
```diff
@@ -786,6 +813,14 @@ def test_create_edge_device_config(
             device_config.hostname if device_config.hostname else "{{HOSTNAME}}"
         )
 
+        # respect existing config reprovisioning mode
+        if device_config_path:
+            with open(device_config_path, "rb") as file:
+                expected_config = tomli.load(file)
+                assert device_toml["auto_reprovisioning_mode"] == getattr(expected_config, "auto_reprovisioning_mode", "Dynamic")
```
Comment (digimaun) [inline azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py]: The test case with the device_config_path is pointing to [this one](https://github.com/Azure/azure-iot-cli-extension/blob/dev/azext_iot/tests/iothub/devices/device_configs/device_config.toml) , right? In that case when is auto_reprovisioning_mode not Dynamic?  I don't see the `auto_reprovisioning_mode` field set in the file.

Comment (c-ryan-k) [reply azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py]: It's actually changed in this PR - [I've set the test `config.toml` to use `auto_reprovisioning_mode = "OnErrorOnly"`](https://github.com/Azure/azure-iot-cli-extension/pull/613/files#diff-7ca392c8d88a5cc372d64ce3e35b9ede99a5ef0f7951edd65f264e9bf4808cf1R6)

Comment (c-ryan-k) [reply azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py]: And just for clarity - this test file is not the one used as the basis / default template if one isn't provided, that's all kept in  the `edge_device_config.py` [here](https://github.com/Azure/azure-iot-cli-extension/pull/613/files#diff-b9fea65352456a520c7b7f54f2dd74b165bf4b1120b7f12e07f5749a0c1807e9R106)

Comment (digimaun) [reply azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py]: Missed that thanks....do we need the "Dynamic" fallback in `getattr`?

Edit: Talking about the line 820 assertion here

Comment (c-ryan-k) [reply azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py]: I believe so - I want to check that:
- If there is an existing mode provided in the config, ensure that is in the resultant config
- If there isn't, ensure it gets set to "Dynamic"

To be more explicit, I could change it to:
```suggestion
                if not getattr(expected_config, "auto_reprovisioning_mode", None):
                    assert device_toml["auto_reprovisioning_mode"] == "Dynamic"
                else:
                    assert device_toml["auto_reprovisioning_mode"] == expected_config["auto_reprovisioning_mode"]
```

I see what you mean, it kind of hides (even if the result is the same) whether the expected config had no value or a value of "Dynamic"

Comment (digimaun) [reply azext_iot/tests/iothub/devices/test_iot_edge_devices_unit.py]: Yea that's what I was getting at, potentially impacting the intent of the test, I like the more explicit refactor

**Review by digimaun — APPROVED** [review]

**Pull request merged**